sql/src/main/java/org/apache/druid/sql/http/SqlResource.java

```diff
@@ -20,86 +20,160 @@
   private final ServerConfig serverConfig;
   private final ResponseContextConfig responseContextConfig;
   private final DruidNode selfNode;
+  private final SqlLifecycleManager sqlLifecycleManager;
+  private final Map<String, SqlEngine> engines;
 
+  @VisibleForTesting
   @Inject
-  protected SqlResource(
+  public SqlResource(
       final ObjectMapper jsonMapper,
       final AuthorizerMapper authorizerMapper,
-      final @NativeQuery SqlStatementFactory sqlStatementFactory,
-      final SqlLifecycleManager sqlLifecycleManager,
       final ServerConfig serverConfig,
+      final SqlLifecycleManager sqlLifecycleManager,
+      final Map<String, SqlEngine> engines,
       ResponseContextConfig responseContextConfig,
       @Self DruidNode selfNode
   )
   {
+    this.engines = engines;
     this.jsonMapper = Preconditions.checkNotNull(jsonMapper, "jsonMapper");
     this.authorizerMapper = Preconditions.checkNotNull(authorizerMapper, "authorizerMapper");
-    this.sqlStatementFactory = Preconditions.checkNotNull(sqlStatementFactory, "sqlStatementFactory");
-    this.sqlLifecycleManager = Preconditions.checkNotNull(sqlLifecycleManager, "sqlLifecycleManager");
     this.serverConfig = Preconditions.checkNotNull(serverConfig, "serverConfig");
     this.responseContextConfig = responseContextConfig;
     this.selfNode = selfNode;
+    this.sqlLifecycleManager = Preconditions.checkNotNull(sqlLifecycleManager, "sqlLifecycleManager");
+  }
+
+  @GET
+  @Path("/engines")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSupportedEngines(@Context final HttpServletRequest request)
+  {
+    AuthorizationUtils.setRequestAuthorizationAttributeIfNeeded(request);
+    return Response.ok(new SupportedEnginesResponse(engines.keySet())).build();
+  }
+
+  /**
+   * API to list all running queries, for an engine that supports such listings.
+   *
+   * @param selfOnly if true, return queries running on this server. If false, return queries running on all servers.
+   * @param engine engine name.
+   * @param request  http request.
+   */
+  @GET
+  @Path("/queries")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response doGetRunningQueries(
+      @QueryParam("selfOnly") final String selfOnly,
+      @QueryParam("engine") @Nullable final String engine,
+      @Context final HttpServletRequest request
+  )
+  {
+    final AuthenticationResult authenticationResult = AuthorizationUtils.authenticationResultFromRequest(request);
+    final AuthorizationResult stateReadAccess = AuthorizationUtils.authorizeAllResourceActions(
+        authenticationResult,
+        Collections.singletonList(new ResourceAction(Resource.STATE_RESOURCE, Action.READ)),
+        authorizerMapper
+    );
+
+    SqlEngine sqlEngine = getEngine(engine);
+    if (sqlEngine == null) {
+      return Response.status(Status.BAD_REQUEST).entity("Unsupported engine").build();
+    }
+
+    List<QueryInfo> queries = sqlEngine.getRunningQueries(selfOnly != null);
```

Comment:
note: alternate way could be to pass `stateReadAccess` to `getRunningQueries` and make the other end do the filtering/etc (make`permission` accessible somehow)

this may enable to have an empty `QueryInfo` interface

Reply:
Made this change